lib/pure/json.nim

```diff
@@ -677,7 +677,7 @@ proc pretty*(node: JsonNode, indent = 2): string =
   ## Similar to prettyprint in Python.
   runnableExamples:
     let j = %* {"name": "Isaac", "books": ["Robot Dreams"],
-                "details": {"age": 35, "pi": 3.1415}}
+                "details": {"age": 35, "number": 3.125}}
```

Comment:
We're now redefining the value of `pi` just so some tests might pass? :-1:

Reply:
No, we are not redefining the value of pi. It is the value "number" now. The reason for this change is simple. 3.1415 isn't a valid binary floating point value, the value can only be represented approximately. When I print it on this branch the output will be ``3.1415000000000002``. Then 3.1415 isn't the exact value of π either, it is just an approximation, and a pretty bad one to be honest. But the test relies on this exact precision. Then last but not least the test isn't really about π, it is just about printing a floating point number. So I've changed the number to a value that is precisely representable as binary float. I think that's reasonable.

Reply:
That is simply the problem that using 17 digits sometimes produces unnecessary digits.

The string representation 3.1415000000000002 of ryu and Pythons JSON serialization is both `"3.1415"` because there is no need to specify all the extra digits to represent it.

![image](https://user-images.githubusercontent.com/3620703/73925882-0787ec80-48cf-11ea-9b95-14b1735324a3.png)

Reply:
So, for the following example:

```nim
let a = 3.1415
let b = 0.3

echo a
echo b
```

Before this PR the output is:
```
3.1415
0.3
```

and after this PR the output is:
```
3.1415000000000002
0.29999999999999999
```

Reply:
Maybe we should adopt ryu.

Reply:
Oh yes please!

Reply:
@narimiran no the output of your snippit is the following:
```
3.1415000000000002
0.29999999999999999
true
true
```
@bluenote10 True, using 17 digits sometimes produces unnecessary digits, but that is much better than not enough digits where the number can't be converted back to binary without loss of precision. Also I like the numbers ``3.1415000000000002``, because it provides feedback to the user in the form of "sorry, I really can't represent `3.1415` precisely", which really is a very common mistake by many beginner programmers that aren't taught about floating point numbers.

Reply:
@krux02 I also think that full precision is better than lossy. We only have to keep in mind that this change can increase the size of json serialized data significantly, because the behavior will suddenly apply to many value "short" float values. For instance, a simple 0.1 changes from 3 characters to 19 characters.

Ryu would be really great indeed.

Reply:
> if you rebuild the compiler the output is the following:

You're right, I hadn't rebuilt the compiler when I had run that example.